File: cpp/src/DoorLock.cpp
#include "../include/DoorLock.hpp"
#include <cstdio>

DoorLock::DoorLock(int id, const std::string &name, const std::string &location)
    : Device(id, name, location), isLocked(true) {
    std::printf("DoorLock '%s' инициализирован (id=%d)\n", getName().c_str(), getId());
}

DoorLock::DoorLock(const DoorLock& other)
    : Device(other), isLocked(other.isLocked) {}

DoorLock::~DoorLock() {}

void DoorLock::lock() {
    isLocked = true;
    std::printf("DoorLock::lock вызван для '%s'\n", getName().c_str());
}

void DoorLock::unlock() {
    isLocked = false;
    std::printf("DoorLock::unlock вызван для '%s'\n", getName().c_str());
}

void DoorLock::alertIfForced() {
    std::printf("DoorLock::alertIfForced для '%s' (демо)\n", getName().c_str());
}

void DoorLock::showStatus() {
    std::printf("DoorLock '%s' locked: %s\n", getName().c_str(), isLocked ? "ДА" : "НЕТ");
}

File: java/DoorLock.java
package smarthome;

public class DoorLock extends AbstractDevice {

    private boolean locked;

    public DoorLock(String id, String name, String location, int powerUsage) {
        super(id, name, location, powerUsage);
        locked = true;
    }

    public void lock() {
        locked = true;
        System.out.println(name + ": заблокирован");
    }

    public void unlock() {
        locked = false;
        System.out.println(name + ": разблокирован");
    }

    @Override
    public void perform(String command) {
        try {
            String c = command.toLowerCase();

            if (c.contains("lock") || c.contains("заблок")) {
                lock();
            } else if (c.contains("unlock") || c.contains("разблок")) {
                unlock();
            } else {
                System.out.println(name + ": неизвестная команда для замка");
            }

        } catch (Exception e) {
            System.out.println(name + ": ошибка команды замка: " + e.getMessage());
        }
    }

    @Override
    public String toString() {
        return super.toString() + " locked=" + locked;
    }
}
